// n8n-nodes-base/dist/nodes/RunAt.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RunAt = void 0;
class RunAt {
    constructor() {
        this.description = {
            displayName: 'RunAt',
            name: 'runAt',
            icon: 'fa:calendar',
            group: ['trigger'],
            version: 1,
            description: 'Triggers the workflow at a specific time',
            defaults: {
                name: 'RunAt',
                color: '#00FF00',
            },
            inputs: [],
            outputs: ['main'],
            properties: [
                {
                    displayName: 'Trigger Times',
                    name: 'triggerTimes',
                    type: 'dateTime',
                    typeOptions: {
                        multipleValues: true,
                        multipleValueButtonText: 'Add Time',
                    },
                    default: '',
                    description: 'Triggers for the workflow',
                    placeholder: 'Add Time',
                },
            ],
        };
    }
    async trigger() {
        return {};
    }
}
exports.RunAt = RunAt;
