// n8n-nodes-base/dist/nodes/Slack/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateJSON = exports.slackApiRequestAllItems = exports.slackApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const _ = require("lodash");
async function slackApiRequest(method, resource, body = {}, query = {}, headers = undefined, option = {}) {
    const authenticationMethod = this.getNodeParameter('authentication', 0, 'accessToken');
    let options = {
        method,
        headers: headers || {
            'Content-Type': 'application/json; charset=utf-8',
        },
        body,
        qs: query,
        uri: `https://slack.com/api${resource}`,
        json: true,
    };
    options = Object.assign({}, options, option);
    if (Object.keys(body).length === 0) {
        delete options.body;
    }
    if (Object.keys(query).length === 0) {
        delete options.qs;
    }
    try {
        let response;
        if (authenticationMethod === 'accessToken') {
            const credentials = this.getCredentials('slackApi');
            if (credentials === undefined) {
                throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'No credentials got returned!');
            }
            options.headers.Authorization = `Bearer ${credentials.accessToken}`;
            response = await this.helpers.request(options);
        }
        else {
            const oAuth2Options = {
                tokenType: 'Bearer',
                property: 'authed_user.access_token',
            };
            response = await this.helpers.requestOAuth2.call(this, 'slackOAuth2Api', options, oAuth2Options);
        }
        if (response.ok === false) {
            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Slack error response: ' + JSON.stringify(response));
        }
        return response;
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.slackApiRequest = slackApiRequest;
async function slackApiRequestAllItems(propertyName, method, endpoint, body = {}, query = {}) {
    const returnData = [];
    let responseData;
    query.page = 1;
    query.count = 100;
    do {
        responseData = await slackApiRequest.call(this, method, endpoint, body, query);
        query.cursor = encodeURIComponent(_.get(responseData, 'response_metadata.next_cursor'));
        query.page++;
        returnData.push.apply(returnData, responseData[propertyName]);
    } while ((responseData.response_metadata !== undefined &&
        responseData.response_metadata.mext_cursor !== undefined &&
        responseData.response_metadata.next_cursor !== '' &&
        responseData.response_metadata.next_cursor !== null) ||
        (responseData.paging !== undefined &&
            responseData.paging.pages !== undefined &&
            responseData.paging.page !== undefined &&
            responseData.paging.page < responseData.paging.pages));
    return returnData;
}
exports.slackApiRequestAllItems = slackApiRequestAllItems;
function validateJSON(json) {
    let result;
    try {
        result = JSON.parse(json);
    }
    catch (exception) {
        result = undefined;
    }
    return result;
}
exports.validateJSON = validateJSON;
