// n8n-nodes-base/dist/nodes/Facebook/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.facebookApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function facebookApiRequest(method, resource, body = {}, qs = {}, uri, option = {}) {
    let credentials;
    if (this.getNode().name.includes('Trigger')) {
        credentials = this.getCredentials('facebookGraphAppApi');
    }
    else {
        credentials = this.getCredentials('facebookGraphApi');
    }
    qs.access_token = credentials.accessToken;
    const options = {
        headers: {
            accept: 'application/json,text/*;q=0.99',
        },
        method,
        qs,
        body,
        gzip: true,
        uri: uri || `https://graph.facebook.com/v8.0${resource}`,
        json: true,
    };
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.facebookApiRequest = facebookApiRequest;
