// n8n-nodes-base/dist/nodes/Mautic/ContactCompanyDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.contactCompanyFields = exports.contactCompanyOperations = void 0;
exports.contactCompanyOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        displayOptions: {
            show: {
                resource: [
                    'contactCompany',
                ],
            },
        },
        options: [
            {
                name: 'Add',
                value: 'add',
                description: 'Add contact to a company',
            },
            {
                name: 'Remove',
                value: 'remove',
                description: 'Remove a contact from a company',
            },
        ],
        default: 'create',
        description: 'The operation to perform.',
    },
];
exports.contactCompanyFields = [
    {
        displayName: 'Contact ID',
        name: 'contactId',
        type: 'string',
        displayOptions: {
            show: {
                resource: [
                    'contactCompany',
                ],
                operation: [
                    'add',
                    'remove',
                ],
            },
        },
        default: '',
        description: 'The ID of the contact.',
    },
    {
        displayName: 'Company ID',
        name: 'companyId',
        type: 'string',
        displayOptions: {
            show: {
                resource: [
                    'contactCompany',
                ],
                operation: [
                    'add',
                    'remove',
                ],
            },
        },
        default: '',
        description: 'The ID of the company.',
    },
];
