// n8n-nodes-base/dist/credentials/SalesforceOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SalesforceOAuth2Api = void 0;
class SalesforceOAuth2Api {
    constructor() {
        this.name = 'salesforceOAuth2Api';
        this.extends = [
            'oAuth2Api',
        ];
        this.displayName = 'Salesforce OAuth2 API';
        this.documentationUrl = 'salesforce';
        this.properties = [
            {
                displayName: 'Authorization URL',
                name: 'authUrl',
                type: 'hidden',
                default: 'https://login.salesforce.com/services/oauth2/authorize',
                required: true,
            },
            {
                displayName: 'Access Token URL',
                name: 'accessTokenUrl',
                type: 'string',
                default: 'https://yourcompany.salesforce.com/services/oauth2/token',
                required: true,
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: 'full refresh_token',
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: '',
            },
            {
                displayName: 'Authentication',
                name: 'authentication',
                type: 'hidden',
                default: 'header',
                description: 'Method of authentication.',
            },
        ];
    }
}
exports.SalesforceOAuth2Api = SalesforceOAuth2Api;
