// n8n-nodes-base/dist/nodes/AgileCrm/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateJSON = exports.agileCrmApiRequestUpdate = exports.agileCrmApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function agileCrmApiRequest(method, endpoint, body = {}, query = {}, uri) {
    const credentials = this.getCredentials('agileCrmApi');
    const options = {
        method,
        headers: {
            'Accept': 'application/json',
        },
        auth: {
            username: credentials.email,
            password: credentials.apiKey,
        },
        uri: uri || `https://${credentials.subdomain}.agilecrm.com/dev/${endpoint}`,
        json: true,
    };
    if (method !== 'GET' && method !== 'DELETE') {
        options.body = body;
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.agileCrmApiRequest = agileCrmApiRequest;
async function agileCrmApiRequestUpdate(method = 'PUT', endpoint, body = {}, query = {}, uri) {
    var _a, _b;
    const credentials = this.getCredentials('agileCrmApi');
    const baseUri = `https://${credentials.subdomain}.agilecrm.com/dev/`;
    const options = {
        method,
        headers: {
            'Accept': 'application/json',
        },
        body: { id: body.id },
        auth: {
            username: credentials.email,
            password: credentials.apiKey,
        },
        uri: uri || baseUri,
        json: true,
    };
    const successfulUpdates = [];
    let lastSuccesfulUpdateReturn;
    const payload = body;
    try {
        if (payload.properties) {
            options.body.properties = payload.properties;
            options.uri = baseUri + 'api/contacts/edit-properties';
            lastSuccesfulUpdateReturn = await this.helpers.request(options);
            (_a = payload.properties) === null || _a === void 0 ? void 0 : _a.map((property) => {
                successfulUpdates.push(`${property.name}`);
            });
            delete options.body.properties;
        }
        if (payload.lead_score) {
            options.body.lead_score = payload.lead_score;
            options.uri = baseUri + 'api/contacts/edit/lead-score';
            lastSuccesfulUpdateReturn = await this.helpers.request(options);
            successfulUpdates.push('lead_score');
            delete options.body.lead_score;
        }
        if (body.tags) {
            options.body.tags = payload.tags;
            options.uri = baseUri + 'api/contacts/edit/tags';
            lastSuccesfulUpdateReturn = await this.helpers.request(options);
            (_b = payload.tags) === null || _b === void 0 ? void 0 : _b.map((tag) => {
                successfulUpdates.push(`(Tag) ${tag}`);
            });
            delete options.body.tags;
        }
        if (body.star_value) {
            options.body.star_value = payload.star_value;
            options.uri = baseUri + 'api/contacts/edit/add-star';
            lastSuccesfulUpdateReturn = await this.helpers.request(options);
            successfulUpdates.push('star_value');
            delete options.body.star_value;
        }
        return lastSuccesfulUpdateReturn;
    }
    catch (error) {
        if (successfulUpdates.length === 0) {
            throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
        }
        else {
            throw new n8n_workflow_1.NodeApiError(this.getNode(), error, { message: `Not all properties updated. Updated properties: ${successfulUpdates.join(', ')}`, description: error.message, httpCode: error.statusCode });
        }
    }
}
exports.agileCrmApiRequestUpdate = agileCrmApiRequestUpdate;
function validateJSON(json) {
    let result;
    try {
        result = JSON.parse(json);
    }
    catch (exception) {
        result = undefined;
    }
    return result;
}
exports.validateJSON = validateJSON;
